### PR TITLE
Support User: Perform a complete page refresh when activating

### LINKED ISSUE #3029: Support User: Sidebar disappears when switching
Reported by @dllh in #2975

pasted_image_3_02_2016__1_10_pm

### LINKED ISSUE #3067: Support User: Cannot read property 'jetpack' of undefined when switching users
Error Message:

> Uncaught TypeError: Cannot read property 'jetpack' of undefined

It looks like this is caused by cached `recentlySelected` sites - a site which doesn't belong to the user sticks in the list as `undefined`

### LINKED ISSUE #3068: Support User: Original site is shown when switching users
When switching users, the original selected site is still shown in the sidebar.

This also appears to be causing the following errors:
screen shot 2016-02-04 at 5 17 30 pm

### PR DESCRIPTION
**Pull request by jordwest**
This performs a full page reload when a support user token is activated/deactivated to ensure all user data is fully refreshed. Between refreshes, the support token is stored in localstorage using the **store** module. On the following boot of Calypso, the User initialization sequence is intercepted before the original user is loaded, to ensure none of the original user's data is loaded.

Additionally, an error message is now displayed when an error occurs while fetching a token.

No other functionality has been changed; the testing procedure is the same (ref: p4TIVU-2OW-p2)

A few notes for review:
 - The Redux store wasn't an option for storing the support token, as we need to check for the token at the earliest point during **User** module initialization, before the Redux store is initialized.
 - Attempts to use **localforage** also caused issues due to its async API and the synchronous nature of the boot process - see #3556

Fixes #3029, fixes #3067, fixes #3068 as well as various other non-reported issues (ref p4TIVU-2OW-p2)

cc @dllh @gwwar @mtias

### CODE REVIEW

**Comment by dllh** [comment]
I gave this a test drive, and it's working well for me functionally.

File: client/layout/index.jsx [hunk 1]
```diff
@@ -179,7 +178,7 @@ export default connect(
 		const { isLoading, section, hasSidebar, isFullScreen, chunkName } = state.ui;
 		return {
 			isLoading,
-			isSupportUser: isSupportUser( state ),
+			isSupportUser: state.support.isSupportUser,
```
Comment (gwwar) [inline client/layout/index.jsx]: Were the selector imports becoming too lengthy?

File: client/support/support-user/index.jsx [hunk 2]
```diff
@@ -53,16 +50,17 @@ const SupportUser = React.createClass( {
...
 
 const mapDispatchToProps = ( dispatch ) => {
 	return {
-		supportUserTokenFetch: flowRight( dispatch, supportUserTokenFetch ),
-		supportUserRestore: flowRight( dispatch, supportUserRestore ),
+		supportUserTokenFetch: supportUser.fetchToken.bind( supportUser ),
+		supportUserRestore: supportUser.rebootNormally,
 		supportUserToggleDialog: flowRight( dispatch, supportUserToggleDialog ),
 	}
 }
```
Comment (gwwar) [inline client/support/support-user/index.jsx]: Missing ;

File: client/support/support-user/index.jsx [hunk 3]
```diff
@@ -53,16 +50,17 @@ const SupportUser = React.createClass( {
 
 const mapStateToProps = ( state ) => {
 	return {
-		isSupportUser: isSupportUser( state ),
+		isSupportUser: state.support.isSupportUser,
 		isTransitioning: state.support.isTransitioning,
 		showDialog: state.support.showDialog,
+		errorMessage: state.support.errorMessage,
 	}
 }
```
Comment (gwwar) [inline client/support/support-user/index.jsx]: Missing ;

File: client/support/support-user/index.jsx [hunk 4]
```diff
@@ -53,16 +50,17 @@ const SupportUser = React.createClass( {
...
 
 const mapDispatchToProps = ( dispatch ) => {
 	return {
-		supportUserTokenFetch: flowRight( dispatch, supportUserTokenFetch ),
-		supportUserRestore: flowRight( dispatch, supportUserRestore ),
+		supportUserTokenFetch: supportUser.fetchToken.bind( supportUser ),
+		supportUserRestore: supportUser.rebootNormally,
```
Comment (gwwar) [inline client/support/support-user/index.jsx]: This feels a bit weird to be mapping this to class instance methods.

File: client/lib/user/support-user-interop.js [hunk 5]
```diff
@@ -18,27 +24,117 @@ import { getSupportUser, getSupportToken } from 'state/support/selectors';
...
+	/**
+	  * Reboot Calypso as the support user
+	  * @param  {string} user  The support user's username
+	  * @param  {string} token The support token
+	  */
+	rebootWithToken( user, token ) {
+		debug( 'Rebooting Calypso with support user' );
+
+		store.set( STORAGE_KEY, { user, token } );
+		window.location.reload();
```
Comment (gwwar) [inline client/lib/user/support-user-interop.js]: While the features are still in progress, we probably need to clear out the sync-handler (cc @retrofox) and any persisted redux state in addition to the page reload.

`User.prototype.clearStoreIfChanged` in `lib/user` should be updated to handle those localforage keys.

File: client/lib/user/support-user-interop.js [hunk 6]
```diff
@@ -18,27 +24,117 @@ import { getSupportUser, getSupportToken } from 'state/support/selectors';
...
-		const state = reduxStore.getState();
+	/**
+	 * Reboot normally as the main user
+	 */
+	rebootNormally() {
+		debug( 'Rebooting Calypso normally' );
 
-		if ( wpcom.setSupportUserToken( getSupportUser( state ), getSupportToken( state ) ) ) {
-			onTokenChange();
+		store.remove( STORAGE_KEY );
```
Comment (gwwar) [inline client/lib/user/support-user-interop.js]: In a future PR we can work around the sync requirement.

File: client/support/support-user/index.jsx (hunk 4 above)
Comment (jordwest) [inline client/support/support-user/index.jsx]: Previously this selector was using the support user and token in the Redux store to determine whether support user was active. Now that we're storing those outside the Redux store, I've opted for just a boolean flag in the Redux state to keep track of whether support user is active or not. We could keep the selector, but it would just be:

```js
const isSupportUser = ( state ) => state.support.isSupportUser
```

File: client/layout/index.jsx (hunk 1 above)
Comment (jordwest) [inline client/layout/index.jsx]: Previously this selector was using the support user and token in the Redux store to determine whether support user was active. Now that we're storing those outside the Redux store, I've opted for just a boolean flag in the Redux state to keep track of whether support user is active or not. We could keep the selector, but it would just be:

```js
const isSupportUser = ( state ) => state.support.isSupportUser
```

File: client/lib/user/support-user-interop.js (hunk 5 above)
Comment (jordwest) [inline client/lib/user/support-user-interop.js]: Could we reset relevant parts of the redux state tree when a `CURRENT_USER_ID_SET` action is dispatched?

Comment (gwwar) [inline client/lib/user/support-user-interop.js]: Good idea. @aduth can you think of any other cases we might run into besides login/logout/switch user where current user id is updated?

File: client/lib/user/support-user-interop.js [hunk 7]
```diff
@@ -18,27 +24,117 @@ import { getSupportUser, getSupportToken } from 'state/support/selectors';
  *
  * @param {Object}  reduxStore  The global redux store instance
  */
-export default function( reduxStore ) {
-	const user = new User();
+class SupportUser {
```
Comment (gwwar) [inline client/lib/user/support-user-interop.js]: We can maybe leave this for a future PR, but since this is basically a singleton, I think we may be able to work around needing a class. Maybe something like:
```javascript
const reduxStoreReadyPromise = new Promise( ( resolve ) => { 
    setReduxStore = ( store ) => resolve( store ); 
} );
const supportUserIsEnabled = config.isEnabled( 'support-user' );
const export fetchToken( user, password ) {
    if( supportUserIsEnabled ) { 
        return reduxStoreReadyPromise.then(
             //...
        );
    }
}
```

File: client/lib/user/support-user-interop.js (hunk 5 above)
Comment (gwwar) [inline client/lib/user/support-user-interop.js]: @jordwest how about we spin up a separate PR for this one. We have a bit of time before `sync-handler `or `persist-redux` are turned on in prod.

**Comment by gwwar** [comment]
:+1: Nice work! :sparkles: It's really coming along.

One thing we may want to clear up in this PR or in an immediate follow up is that the currentUser id doesn't change back to the original user id after exiting the mode.

As is, I think it's good to :ship: as an outstanding wip.

File: client/lib/user/support-user-interop.js (hunk 5 above)
Comment (jordwest) [inline client/lib/user/support-user-interop.js]: Actually, this raises another question. Would we even want to clear out the state tree on `CURRENT_USER_ID_SET`? I believe it happens on each page reload (not just login) - so we don't want to clear out the persisted data if the user ID hasn't changed. Maybe there's room for a `CURRENT_USER_ID_CHANGED` action

**Comment by jordwest** [comment]
Thanks for the review @gwwar. I've fixed up the typos and will create issues to track the other remaining bits.

For the time being I've fixed the issue with currentUser id not changing back to the original user by just nuking the store completely (as is done in `User` when it detects a change):
```diff
 	rebootNormally() {
 		debug( 'Rebooting Calypso normally' );

-		store.remove( STORAGE_KEY );
+		store.clear();
 		window.location.reload();
 	}
```

If there are no other objections I'll merge shortly

**Comment by gwwar** [comment]
As long as you're awake for the deploy, go for it. :smile:

File: client/lib/user/support-user-interop.js (hunk 7 above)
Comment (jordwest) [inline client/lib/user/support-user-interop.js]: moved to :arrow_right:  #3587

File: client/lib/user/support-user-interop.js (hunk 5 above)
Comment (jordwest) [inline client/lib/user/support-user-interop.js]: Tracking this in #3586

Comment (retrofox) [inline client/lib/user/support-user-interop.js]: > While the features are still in progress, we probably need to clear out the sync-handler (cc @retrofox) and any persisted redux state in addition to the page reload.

To clean whole sync-handler data we can use `syncHandler.clearAll()`. I was thinking to keep some `public` data for instance public blogs but maybe it doesn't make sense.
